File: app/utils/BufferedCsvWriter.ts
import RNFS from 'react-native-fs';

export class BufferedCsvWriter {
    private buffer: string[] = [];
    private filePath: string;
    private bufferSize: number;
    private initialized = false;

    constructor(filePath: string, bufferSize: number = 1000) {
        this.filePath = filePath;
        this.bufferSize = bufferSize;
    }

    async initialize(header: string = 'iteration,executionTimeMs,details,intervalStartMs,intervalDurationMs,cumulativeTimeMs') {
        if (this.initialized) return;
        try {
            // Verify path is writable. usually RNFS.DocumentDirectoryPath is used.
            const dir = this.filePath.substring(0, this.filePath.lastIndexOf('/'));
            await RNFS.mkdir(dir);
            await RNFS.writeFile(this.filePath, header + '\n', 'utf8');
            this.initialized = true;
        } catch (e) {
            console.error('Failed to init CSV writer', e);
        }
    }

    async write(
        iteration: number,
        executionTimeMs: number,
        details: string,
        intervalStartMs: number,
        intervalDurationMs: number,
        cumulativeTimeMs: number
    ) {
        const escapedDetails = details.replace(/"/g, '""');
        const needsQuote = escapedDetails.includes(',') || escapedDetails.includes('\n');
        const finalDetails = needsQuote ? `"${escapedDetails}"` : escapedDetails;

        const line = `${iteration},${executionTimeMs},${finalDetails},${intervalStartMs},${intervalDurationMs},${cumulativeTimeMs}`;
        this.buffer.push(line);

        if (this.buffer.length >= this.bufferSize) {
            await this.flush();
        }
    }

    async flush() {
        if (this.buffer.length === 0) return;

        try {
            const chunk = this.buffer.join('\n') + '\n';
            this.buffer = [];
            await RNFS.appendFile(this.filePath, chunk, 'utf8');
        } catch (e) {
            console.error('Flush error', e);
            // Don't clear buffer on error to retry? Or drop to avoid OOM?
            // For now, we drop to avoid getting stuck, but log error.
        }
    }

    async close() {
        await this.flush();
    }
}

File: flutter/lib/utils/buffered_csv_writer.dart
import 'dart:io';

class BufferedCsvWriter {
  final String filePath;
  final int bufferSize;
  final String testName;
  final List<String> _buffer = [];
  bool _initialized = false;
  IOSink? _sink;

  BufferedCsvWriter(this.filePath, {this.bufferSize = 1000, this.testName = 'unknown'});

  Future<void> initialize({String header = 'iteration,execution_time_ms,details,interval_start_ms,interval_duration_ms,cumulative_time_ms'}) async {
    if (_initialized) return;
    try {
      final file = File(filePath);
      final parent = file.parent;
      if (!await parent.exists()) {
        await parent.create(recursive: true);
      }
      // Use append mode for safety, but initially write header if new
      _sink = file.openWrite(mode: FileMode.append);
      
      if (await file.length() == 0) {
        _sink?.writeln(header);
        await _sink?.flush(); // Ensure header is written immediately
      }
      _initialized = true;
    } catch (e) {
      print('Failed to init CSV writer: $e');
    }
  }

  Future<void> write(
    int iteration,
    int executionTimeMs,
    String details, {
    int intervalStartMs = 0,
    int intervalDurationMs = 0,
    int cumulativeTimeMs = 0,
  }) async {
    if (!_initialized) return;

    final safeDetails = csvEscape(details);
    final line = '$iteration,$executionTimeMs,$safeDetails,$intervalStartMs,$intervalDurationMs,$cumulativeTimeMs';
    _buffer.add(line);

    if (_buffer.length >= bufferSize) {
      await flush();
    }
  }

  Future<void> flush() async {
    if (_sink == null) return;
    try {
      if (_buffer.isNotEmpty) {
        final chunk = _buffer.join('\n');
        _sink?.writeln(chunk);
        _buffer.clear();
      }
      // Always flush the sink to ensure data is written to disk
      await _sink?.flush();
    } catch (e) {
      print('Flush error: $e');
    }
  }

  Future<void> close() async {
    await flush();
    await _sink?.close();
    _initialized = false;
  }

  String csvEscape(String v) {
    if (v.contains(',') || v.contains('"') || v.contains('\n')) {
      return '"${v.replaceAll('"', '""')}"';
    }
    return v;
  }
}

